README
NAME
Mojolicious::Plugin::Authorization - A plugin to make Authorization a bit easier

VERSION
version 1.03

SYNOPSIS
    use Mojolicious::Plugin::Authorization
    $self->plugin('Authorization' => {
        'has_priv'   => sub { ... },
        'is_role'    => sub { ... },
        'user_privs' => sub { ... },
        'user_role'  => sub { ... },
    });
    if ($self->has('delete_all', { optional => 'extra data stuff' })) {
        ...
    }

DESCRIPTION
A very simple API implementation of role-based access control (RBAC). This plugin is only an API you will
have to do all the work of setting up your roles and privileges and then provide four subs that are used by
the plugin.
The plugin expects that the current session will be used to get the role its privileges. It also assumes that
you have already been authenticated and your role set.
That is about it you are free to implement any system you like.

METHODS

has_priv('privilege', $extra_data) or has_privilege('privilege', $extra_data)

'has_priv' and 'has_privilege' will use the supplied "has_priv" subroutine ref to check if the current session has the
given privilege. It should return true when the session has the privilege or false otherwise.
You can pass additional data along in the extra_data hashref, it will be passed to your "has_priv"
subroutine as-is.

is('role',$extra_data)

'is' will use the supplied "is_role" subroutine ref to check if the current session is the
given role. I should return true when the session has privilege or false otherwise.
You can pass additional data along in the extra_data hashref, it will be passed to your "is_role"
subroutine as-is.

privileges($extra_data)

'privileges' will use the supplied "user_privs" subroutine ref and return the privileges of the current session.
You can pass additional data along in the extra_data hashref, it will be passed to your "user_privs"
subroutine as-is. The returned data is dependant on the supplied "user_privs" subroutine.

role($extra_data)

'role' will use the supplied "user_role" subroutine ref and return the role of the current session.
You can pass additional data along in the extra_data hashref, it will be passed to your "user_role"
subroutine as-is. The returned data is dependant on the supplied "user_privs" subroutine.

CONFIGURATION

The following options must be set for the plugin:

    has_priv (REQUIRED) A coderef for checking to see if the current session has a privilege (see "HAS_PRIV").
    is_role (REQUIRED) A coderef for checking to see if the current session is a certain role (see "IS").
    user_privs (REQUIRED) A coderef for returning the privileges of the current session (see "PRIVILEGES").
    user_role (REQUIRED) A coderef for retiring the role of the current session (see "ROLE").

HAS_PRIV

Has_priv is used when you need to confirm that the current session has the given privilege.
The coderef you pass to the "has_priv" configuration key has the following signature:
    sub {
        my ($app, $privilege,$extradata) = @_;
        ...
    }
You must return either 0 for a fail and 1 for a pass.  This allows "ROUTING VIA CONDITION" to work correctly.

IS

Is is used when you need to confirm that the current session is set to the given role.
The coderef you pass to the "is_role" configuration key has the following signature:
    sub {
        my ($app, $role, $extradata) = @_;
        ...
        return $role;
    }
You must return either 0 for a fail and 1 for a pass.  This allows "ROUTING VIA CONDITION" to work correctly.

PRIVILEGES

Privileges is used when you need to get all the privileges of the current session.
The coderef you pass to the "user_privs" configuration key has the following signature:
    sub {
        my ($app,$extradata) = @_;
        ...
        return $privileges;
    }
You can return anything you want. It would normally be an arrayref of privileges but you are free to
return a scalar, hashref, arrayref, blessed object, or undef.

ROLE

Role is used when you need to get the role of the current session.
The coderef you pass to the "user_role" configuration key has the following signature:
    sub {
        my ($app,$extradata) = @_;
        ...
        return $role;
    }
You can return anything you want. It would normally be just a scalar but you are free to
return a scalar, hashref, arrayref, blessed object, or undef.

EXAMPLES

For a code example using this, see the t/01-functional.t test,
it uses Mojolicious::Lite and this plugin.

ROUTING VIA CONDITION
This plugin also exports a routing condition you can use in order to limit access to certain documents to only
sessions that have a privilege.

    $r->route('/delete_all')->over(has_priv => 'delete_all')->to('mycontroller#delete_all');
    my $delete_all_only = $r->route('/members')->over(has_priv => 'delete_all')->to('members#delete_all');
    $delete_all_only->route('delete')->to('members#delete_all');

If the session does not have the 'delete_all' privilege, these routes will not be considered by the dispatcher and unless you have set up a catch-all route,
 a 404 Not Found will be generated instead.
Another condition you can use to limit access to certain documents to only those sessions that
have a role.

    $r->route('/view_all')->over(is => 'ADMIN')->to('mycontroller#view_all');
    my $view_all_only = $r->route('/members')->over(is => 'view_all')->to('members#view_all');
    $view_all_only->route('view')->to('members#view_all');

If the session is not the 'ADMIN' role, these routes will not be considered by the dispatcher and unless you have set up a catch-all route,
 a 404 Not Found will be generated instead.
This behavior is similar to the "has" condition.

ROUTING VIA CALLBACK
It is not recommended to route un-authorized requests to anything but a 404 page. If you do route to some sort
of 'You are not allowed page' you are telling a hacker that the URL was correct while the 404 tells them nothing.
This is just my opinion.

SEE ALSO

Mojolicious::Sessions, Mojocast 3: Authorization <http://mojocasts.com/e3#>

AUTHOR

John Scoles, "<byterock  at hotmail.com>"

BUGS / CONTRIBUTING

Please report any bugs or feature requests through the web interface at <https://github.com/byterock/mojolicious-plugin-authorization/issues>.

SUPPORT

You can find documentation for this module with the perldoc command.
    perldoc Mojolicious::Plugin::Authorization

You can also look for information at:

    * AnnoCPAN: Annotated CPAN documentation
      <http://annocpan.org/dist/Mojolicious-Plugin-Authorization>
    * CPAN Ratings
      <http://cpanratings.perl.org/d/Mojolicious-Plugin-Authorization>
    * Search CPAN
      <http://search.cpan.org/dist/Mojolicious-Plugin-Authorization/>

ACKNOWLEDGEMENTS

Ben van Staveren   (madcat)
    -   For 'Mojolicious::Plugin::Authentication' which I used as a guide in writing up this one.
Chuck Finley
    -   For staring me off on this.
Abhijit Menon-Sen
    -   For the routing suggestions
Roland Lammel
    -   For some other good suggestions

LICENSE AND COPYRIGHT

Copyright 2012 John Scoles.
This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.
See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for Mojolicious-Plugin-Authorization

Changes in 1.01
    Initail version
Changes in 1.02
    Fixed the pod
Changes in 1.03
    3'rd Times a charm

Build.PL
use strict;
use warnings;
use Module::Build 0.38;
my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.38",
    "Mojolicious::Lite" => 0,
    "Test::Mojo" => 0,
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.38"
  },
  "dist_abstract" => "A plugin to make authorization a bit easier",
  "dist_author" => [
    "John Scoles <byterock\@hotmail.com>"
  ],
  "dist_name" => "Mojolicious-Plugin-Authorization",
  "dist_version" => "1.02",
  "license" => "perl",
  "module_name" => "Mojolicious::Plugin::Authorization",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Mojo::Base" => 0
  },
  "script_files" => []
);
my $build = Module::Build->new(%module_build_args);
$build->create_build_script;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker 6.30;
my %WriteMakefileArgs = (
  "ABSTRACT" => "A plugin to make authorization a bit easier",
  "AUTHOR" => "John Scoles <byterock\@hotmail.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.38",
    "Mojolicious::Lite" => 0,
    "Test::Mojo" => 0,
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.38"
  },
  "DISTNAME" => "Mojolicious-Plugin-Authorization",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Mojolicious::Plugin::Authorization",
  "PREREQ_PM" => {
    "Mojo::Base" => 0
  },
  "VERSION" => "1.03",
  "test" => {
    "TESTS" => "t/*.t"
  }
);
unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}
delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };
WriteMakefile(%WriteMakefileArgs);


example/showoff-authorization.pl
#!/usr/bin/env perl
use strict;
use warnings;
use warnings FATAL => qw{ uninitialized };
use autodie;
# Disable IPv6, epoll and kqueue
BEGIN { $ENV{MOJO_NO_IPV6} = $ENV{MOJO_POLL} = 1 }
use Mojolicious::Lite;
=pod
=head1 Title
  showoff-authorization.pl --- an example of the Mojolicious::Plugin::Authorization module by John Scoles
=head1 Invocation
  $ perl showoff-authorization.pl daemon
=head1 Notes
My first crack at a Mojo plugin a steal from Ben van Staveren's Authentication so I owe him and some others
a great note of thanks
Like Authentication this is a very a simple application. It supplies the framwork and you have to give it
the guts which this little progam shows.
I did not add in any Authentication as that is up to you to build. In this test I just assume you are
authenticated on the session ans that session has a role hash on it.
=head1 Versions
  0.1: May 01 2012
=cut
################################################################
### miniauthorfile.pm lays out basic functionality for the miniauthorfile
use miniauthorfile;
my $roles = miniauthorfile->new('miniauthorfile.txt');
################################################################
plugin 'authorization', {

  has_priv => sub {
     my $self = shift;
     my ($priv, $extradata) = @_;
     return 0
       unless($self->session('role'));
     my $role  = $self->session('role');
     my $privs = $roles->{$role};
     return 1
       if exists($privs->{$priv});
     return 0;
  },
  is_role => sub {
    my $self = shift;
    my ($role, $extradata) = @_;
    return 0
       unless($self->session('role'));
    return 1
       if ($self->session('role') eq $role);
   return 0;
  },
  user_privs => sub {
    my $self = shift;
    my ($extradata) = @_;
    return []
       unless($self->session('role'));
    my $role  = $self->session('role');
    my $privs = $roles->{$role};
    return keys(%{$privs});
  },
  user_role => sub {
    my $self = shift;
    my ($extradata) = @_;
    return $self->session('role');
  },
 };
################################################################
get '/' => sub {
  my $self = shift;
  unless($self->session('role')){
    $self->session('role'=>'guest');
  }
  $self->render('index');  ## index needs to be named to match '/'
};
get '/dogshow' => sub {
  my $self = shift;
  unless ($self->has_priv('view')) {
     $self->render('index');
  }
  else{
     $self->stash('role_name'=> $self->role());
     $self->render('dogshow');
  }
};
get '/change/:role' => sub {
  my $self = shift;
  my $role =  $self->param('role');
  $roles->set_role($self->session,$role);
  $self->stash('role_name'=> $self->role());
  $self->render('dogshow');
 # $self->render(template);  ## this is called automatically
};
get '/view' => sub {
  my $self = shift;
  unless ($self->has_priv('view')) {
     $self->render('index');
  }
 # $self->render(template);  ## this is called automatically
};
get '/herd' => sub {
  my $self = shift;
  unless ($self->has_priv('herd')) {
     $self->render('not_allowed');
  }
};
get '/judge' => sub {
  my $self = shift;
  $self->render('not_allowed')
    unless ($self->has_priv('judge'));
  $self->render('all_glory')
    if ($self->is("hypnotoad"))
};
############ these two subs can show you what you can do now, based on authenticated status
get '/my_privs/' => sub {
  my $self = shift;
  $self->render('not_allowed')
    unless ($self->session('role'));
  my @privs = $self->privileges();
  $self->stash('privs'=> \@privs);
};
## /condition/authonly exists as a webpage ONLY after authentication
app->secret('All GLORY to the Hypnotoad!!');  # used for cookies and persistence
app->start();
################################################################
__DATA__
@@ index.html.ep
% layout 'default';
% title 'Root';
<h2> Top Index Page</h2>
<p>The purpose of this little web app is to show an example of <a href="http://mojolicio.us/">Mojolicious</a> and its <a href="http://search.cpan.org/~madcat/Mojolicious-Plugin-Authorization/">Mojolicious::Authorization module</a> by John Scoles.</p>
<p>Go to the trials as a <a href="/change/guest">Guest</a>.</p>
<p>Go to the trials as a <a href="/change/dog">Dog</a>.</p>
<p>Go the trials as a <a href="/change/judge">Judge</a>.</p>
<p>Go the trials as The <a href="/change/hypnotoad">Hypnotoad</a>.</p>
@@ dogshow.html.ep
% layout 'default';
% title 'Pan Galatic Sheep Dog Trials';
<p>Welcome "<%= $role_name %>" to the the Pan Galatic Sheep Dog Trials.</p>
<a href="/">Go home</a><br>
<a href="/view">View a Trial</a><br>
<a href="/herd">Herd some Sheep</a><br>
<a href="/judge">Judge a trial</a><br>
<a href="/my_privs">What are my Privleges</a>
@@ view.html.ep
% layout 'default';
% title 'View Trials';
<h1>Enjoy the Trials</h1>
<p>He's good.</p>
<p>But our real compition is the Hypnotoad</p>
@@ herd.html.ep
% layout 'default';
% title 'Herd Some Sheep';
<h1>Heard Some Sheep</h1>
<p>Woof, Bow-Wow</P>
<p>eye-ball</p>
<p>run~chase</p>
<p>put sheep in pen</p>
@@ judge.html.ep
% layout 'default';
% title 'Judge a Dog';
<h1>Judge a Dog</h1>
<p>5.8</p>
<p>5.9</p>
<p>5.8</p>
<p>5.7</p>
and
<p>4.9</p>
from the Russian Judge
@@ my_privs.html.ep
% layout 'default';
% title 'My Privleges';
<h1>Privleges</h1>
%foreach my $priv (@{$privs}) {
 <%==$priv%> <br>
%}
@@ all_glory.html.ep
% layout 'default';
% title 'Judge a Dog';
<h1>Judge a Dog</h1>
<p>And the winner is </p>
<h1><blink>The Hypnotoad</blink></h1>
<p>All Gloy to the Hypnotoad</p>
Clap-Clap-Clap
@@ not_allowed.html.ep
% layout 'default';
% title 'Page Unavailable';
<h1>I am sorry do to interferance from suicide booths on 'Eminiar VII' you cannot get to this page</h1>
@@ layouts/default.html.ep
<!DOCTYPE html>
<html>
  <head>
    <title><%= title %></title>
  </head>
  <body>
    <hr />
    <h1> Mojolicious: <%= $0 %>: <%= title %> </h1>
    <hr />
    <%= content %>
    <hr />
  </body>
</html>


lib/Mojolicious/Plugin/Authorization.pm
package Mojolicious::Plugin::Authorization;
BEGIN {
  $Mojolicious::Plugin::Authorization::VERSION = '1.03';
}
use Mojo::Base 'Mojolicious::Plugin';
# The dog is good, but our real competition is the Hypnotoad.
sub register {
    my ($self, $app, $args) = @_;
    $args ||= {};
    die __PACKAGE__, ": missing 'has_priv' subroutine ref in parameters\n"
        unless $args->{has_priv} && ref($args->{has_priv}) eq 'CODE';
    die __PACKAGE__, ": missing 'is_role' subroutine ref in parameters\n"
        unless $args->{is_role} && ref($args->{is_role}) eq 'CODE';
    die __PACKAGE__, ": missing 'user_privs' subroutine ref in parameters\n"
        unless $args->{user_privs} && ref($args->{user_privs}) eq 'CODE';
    die __PACKAGE__, ": missing 'user_role' subroutine ref in parameters\n"
        unless $args->{user_role} && ref($args->{user_role}) eq 'CODE';
    my $has_priv_cb    = $args->{has_priv};
    my $is_role_cb     = $args->{is_role};
    my $user_privs_cb  = $args->{user_privs};
    my $user_role_cb   = $args->{user_role};
    $app->routes->add_condition(has_priv => sub {
        my ($r, $c, $captures, $priv) = @_;
        return ($priv && $has_priv_cb->($c,$priv)) ? 1 : 0;
    });
    $app->routes->add_condition(is => sub {
        my ($r, $c, $captures, $role) = @_;
        return ($role && $is_role_cb->($c,$role)) ? 1 : 0;
    });
    $app->helper(privileges => sub {
        my ($c, $user, $extradata) = @_;
        return $user_privs_cb->($c, $user, $extradata);
    });
    $app->helper(has_priv => sub {
        my ($c, $priv, $extradata) = @_;
        my $has_priv = $has_priv_cb->($c, $priv, $extradata);
        return $has_priv;
    });
    $app->helper(has_privilege => sub {
        my ($c, $priv, $extradata) = @_;
        my $has_priv = $has_priv_cb->($c, $priv, $extradata);
        return $has_priv;
    });
    $app->helper(is => sub {
        my ($c, $role, $extradata) = @_;
        return $is_role_cb->($c, $role, $extradata);
    });
    $app->helper(role => sub {
       my ($c, $user, $extradata) = @_;
        return $user_role_cb->($c, $user, $extradata);
    });
}
1;
__END__
=pod

=head1 NAME

Mojolicious::Plugin::Authorization - A plugin to make Authorization a bit easier

=head1 VERSION

version 1.03

=head1 SYNOPSIS

    use Mojolicious::Plugin::Authorization
    $self->plugin('Authorization' => {
        'has_priv'   => sub { ... },
        'is_role'    => sub { ... },
        'user_privs' => sub { ... },
        'user_role'  => sub { ... },
    });
    if ($self->has_priv('delete_all', { optional => 'extra data stuff' })) {
        ...
    }

=head1 DESCRIPTION

A very simple API implementation of role-based access control (RBAC). This plugin is only an API you will
have to do all the work of setting up your roles and privileges and then provide four subs that are used by
the plugin.
The plugin expects that the current session will be used to get the role its privileges. It also assumes that
you have already been authenticated and your role set.
That is about it you are free to implement any system you like.

=head1 METHODS

=head2 has_priv('privilege', $extra_data) or has_privilege('privilege', $extra_data)

'hHas_priv'' and ''has_privilege'' will use the supplied C<has_priv> subroutine ref to check if the current session has the
given privilege. Returns true when the session has the privilege or false otherwise.
You can pass additional data along in the extra_data hashref and it will be passed to your C<has_priv>
subroutine as-is.

=head2 is('role',$extra_data)

'is' will use the supplied C<is_role> subroutine ref to check if the current session is the
given role. Returns true when the session has privilege or false otherwise.
You can pass additional data along in the extra_data hashref and it will be passed to your C<is_role>
subroutine as-is.

=head2 privileges($extra_data)

'pPrivileges'' will use the supplied C<user_privs> subroutine ref and return the privileges of the current session.
You can pass additional data along in the extra_data hashref and it will be passed to your C<user_privs>
subroutine as-is. The returned data is dependant on the supplied C<user_privs> subroutine.

=head2 role($extra_data)

'role' will use the supplied C<user_role> subroutine ref and return the role of the current session.
You can pass additional data along in the extra_data hashref and it will be passed to your C<user_role>
subroutine as-is. The returned data is dependant on the supplied C<user_role> subroutine.

=head1 CONFIGURATION

The following options must be set for the plugin:

=over 4

=item has_priv (REQUIRED) A coderef for checking to see if the current session has a privilege (see L</"HAS PRIV">).

=item is_role (REQUIRED) A coderef for checking to see if the current session is a certain role (see L</"IS ROLE">).

=item user_privs (REQUIRED) A coderef for returning the privileges of the current session (see L</"PRIVILEGES">).

=item user_role (REQUIRED) A coderef for retiring the role of the current session (see L</"ROLE">).

=back

=head2 HAS PRIV

'has_priv' is used when you need to confirm that the current session has the given privilege.
The coderef you pass to the C<has_priv> configuration key has the following signature:

    sub {
        my ($app, $privilege,$extradata) = @_;
        ...
    }

You must return either 0 for a fail and 1 for a pass.  This allows C<ROUTING VIA CONDITION> to work correctly.

=head2 IS

'is' is used when you need to confirm that the current session is set to the given role.
The coderef you pass to the C<is_role> configuration key has the following signature:

    sub {
        my ($app, $role, $extradata) = @_;
        ...
        return $role;
    }

You must return either 0 for a fail and 1 for a pass.  This allows C<ROUTING VIA CONDITION> to work correctly.

=head2 PRIVILEGES

'privileges' is used when you need to get all the privileges of the current session.
The coderef you pass to the C<user_privs> configuration key has the following signature:

    sub {
        my ($app,$extradata) = @_;
        ...
        return $privileges;
    }

You can return anything you want. It would normally be an arrayref of privileges but you are free to
return a scalar, hashref, arrayref, blessed object, or undef.

=head2 ROLE

'role' is used when you need to get the role of the current session.
The coderef you pass to the C<user_privs> configuration key has the following signature:

    sub {
        my ($app,$extradata) = @_;
        ...
        return $role;
    }
    
You can return anything you want. It would normally be just a scalar but you are free to
return a scalar, hashref, arrayref, blessed object, or undef.

=head1 EXAMPLES

For a code example using this, see the F<t/01-functional.t> test,
it uses L<Mojolicious::Lite> and this plugin.

=head1 ROUTING VIA CONDITION

This plugin also exports a routing condition you can use in order to limit access to certain documents to only
sessions that have a privilege.

    $r->route('/delete_all')->over(has_priv => 'delete_all')->to('mycontroller#delete_all');
    my $delete_all_only = $r->route('/members')->over(has_priv => 'delete_all')->to('members#delete_all');
    $delete_all_only->route('delete')->to('members#delete_all');

If the session does not have the 'delete_all' privilege, these routes will not be considered by the dispatcher and unless you have set up a catch-all route,
 a 404 Not Found will be generated instead.

Another condition you can use to limit access to certain documents to only those sessions that
have a role.

    $r->route('/view_all')->over(is => 'ADMIN')->to('mycontroller#view_all');
    my $view_all_only = $r->route('/members')->over(is => 'view_all')->to('members#view_all');
    $view_all_only->route('view')->to('members#view_all');

If the session is not the 'ADMIN' role, these routes will not be considered by the dispatcher and unless you have set up a catch-all route,
 a 404 Not Found will be generated instead.
This behavior is similar to the "has" condition.

=head1 ROUTING VIA CALLBACK

It is not recommended to route un-authorized requests to anything but a 404 page. If you do route to some sort
of 'You are not allowed page' you are telling a hacker that the URL was correct while the 404 tells them nothing.
This is just my opinion.

=head1 SEE ALSO

L<Mojolicious::Sessions>, L<Mojocast 3: Authorization|http://mojocasts.com/e3#>

=head1 AUTHOR

John Scoles, C<< <byterock  at hotmail.com> >>

=head1 BUGS / CONTRIBUTING

Please report any bugs or feature requests through the web interface at L<https://github.com/byterock/mojolicious-plugin-authorization/issues>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.
    perldoc Mojolicious::Plugin::Authorization
You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation L<http://annocpan.org/dist/Mojolicious-Plugin-Authorization>

=item * CPAN Ratings L<http://cpanratings.perl.org/d/Mojolicious-Plugin-Authorization>

=item * Search CPAN L<http://search.cpan.org/dist/Mojolicious-Plugin-Authorization/>

=back

=head1 ACKNOWLEDGEMENTS

Ben van Staveren   (madcat)

    -   For 'Mojolicious::Plugin::Authentication' which I used as a guide in writing up this one.
    
Chuck Finley

    -   For staring me off on this.
    
Abhijit Menon-Sen

    -   For the routing suggestions
    
Roland Lammel

    -   For some other good suggestions

=head1 LICENSE AND COPYRIGHT

Copyright 2012 John Scoles.
This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.
See http://dev.perl.org/licenses/ for more information.

=cut